image/templates/helm/shared/templates/_scanner-v4_init.tpl.htpl

```diff
@@ -24,150 +24,112 @@
+          {{- $lookupOut := dict -}}
+          {{- $_ := include "srox.safeLookup" (list $ $lookupOut "v1" "Secret" $.Release.Namespace "scanner-v4-indexer-tls") -}}
+          {{- if not $lookupOut.result -}}
+            {{/* If generate=null and the resource does not exist yet, attempt to create it. */}}
+            {{/* If lookup is not possible (e.g. in the operator), then 'generate' needs to be set correctly. */}}
+            {{- $_ := set $._rox.scannerV4.indexer.serviceTLS "generate" true -}}
+          {{- end -}}
+        {{- end -}}
+      {{- end }}
+      {{ $cryptoSpec := dict "CN" "SCANNER_V4_INDEXER_SERVICE: Scanner V4 Indexer" "dnsBase" "scanner-v4-indexer" }}
+      {{ include "srox.configureCrypto" (list $ "scannerV4.indexer.serviceTLS" $cryptoSpec) }}
     {{- end -}}
-  {{- end -}}
+  {{ end }}
 
-{{- end -}}
+  {{ if get $components "matcher" }}
+    {{ include "srox.configureImage" (list $ $scannerV4Cfg.matcher.image) }}
+    {{- if eq $.Chart.Name "stackrox-central-services" -}}
+      {{/* Only generate certificate when installing central-services. */}}
+      {{- if kindIs "invalid" $._rox.scannerV4.matcher.serviceTLS.generate }}
+        {{/* We need special handling here, because 'generate' will default to 'false' on upgrades. */}}
+        {{/* And in case scanner V4 was not deployed earlier, we need to make sure that these resources are correctly initialized. */}}
+        {{- if $.Release.IsUpgrade -}}
+          {{- $lookupOut := dict -}}
+          {{- $_ := include "srox.safeLookup" (list $ $lookupOut "v1" "Secret" $.Release.Namespace "scanner-v4-matcher-tls") -}}
+          {{- if not $lookupOut.result -}}
+            {{/* If generate=null and the resource does not exist yet, attempt to create it. */}}
+            {{/* If lookup is not possible (e.g. in the operator), then 'generate' needs to be set correctly. */}}
+            {{- $_ := set $._rox.scannerV4.matcher.serviceTLS "generate" true -}}
+          {{- end -}}
+        {{- end }}
+      {{- end }}
+      {{ $cryptoSpec := dict "CN" "SCANNER_V4_MATCHER_SERVICE: Scanner V4 Matcher" "dnsBase" "scanner-v4-matcher" }}
+      {{ include "srox.configureCrypto" (list $ "scannerV4.matcher.serviceTLS" $cryptoSpec) }}
+    {{- end -}}
+  {{ end }}
 
-{{/* Configure images, certificates and passwords as required. */}}
-{{/* Note that for secured-cluster-services we don't configure certificates here, instead they will be distributed at runtime by Sensor and Central. */}}
-{{ if get $components "indexer" }}
-  {{ include "srox.configureImage" (list $ $scannerV4Cfg.indexer.image) }}
-  {{- if eq $.Chart.Name "stackrox-central-services" -}}
-    {{/* Only generate certificate when installing central-services. */}}
-    {{- if kindIs "invalid" $._rox.scannerV4.indexer.serviceTLS.generate }}
+  {{ if or (get $components "indexer") (get $components "matcher") }}
+    {{- if kindIs "invalid" $._rox.scannerV4.db.serviceTLS.generate }}
```

Comment:
I cannot understand which case it covers, could you please explain?

Reply:
Hmm, the code comment is this:

```
{{/* We need special handling here, because 'generate' will default to 'false' on upgrades. */}}
{{/* And in case scanner V4 was not deployed earlier, we need to make sure that these resources are correctly initialized. */}}
```

Let me add a more verbose explanation to that.

So, imagine you have ACS installed with the Helm chart without Scanner V4.
Now you want to add Scanner V4.

There are some things that the Helm chart generates for your convenience like certificates and passwords. The basic approach to decide if such resources should be generated would be to do it when we are in install mode and don't do it when in upgrade mode.       

But this wouldn't suffice for the situation at hand where we would like to add Scanner V4. Because we are actually not installing fresh, we are in upgrade mode! This is the special case we have to deal with here.

Even though we are in upgrade mode we might need to generate these resources. But we do this only when -- as in the example above -- no certificate is provided.